src/workerd/io/io-context.h

```diff
@@ -983,15 +983,22 @@ jsg::PromiseForResult<Func, T, true> IoContext::awaitIoImpl(
             // before the inner promise.
             func = kj::fwd<Func>(func)]
             (ExceptionOr<T>&& exceptionOrT) mutable {
+    struct FuncResultPair {
+      // It's important that `exceptionOrT` is destroyed before `Func`. Lambda captures are
+      // destroyed in unspecified order, so we wrap them in a struct to make it explicit.
+      Func func;
+      ExceptionOr<T> exceptionOrT;
+    };
+
     return run([resolver = kj::mv(resolver),
-                exceptionOrT = kj::mv(exceptionOrT),
-                maybeAsyncContext = kj::mv(maybeAsyncContext),
-                func = kj::fwd<Func>(func)](Worker::Lock& lock) mutable {
+                funcResultPair = FuncResultPair { kj::fwd<Func>(func), kj::mv(exceptionOrT) },
```

Comment:
A comment explaining this briefly would be helpful

Reply:
ha! nevermind... I wrote this then immediately spotted the comment immediately above. Sigh. Need more coffee.